### PR TITLE
Replace Base64 Storage with MinIO Client for Chat Room Attachments

### LINKED ISSUE #3723: NoMongo: Replace Base64 Storage with MinIO Client in ChatRoom.tsx
Problem  
ChatRoom.tsx currently uses Base64 encoding for storing images and media, leading to high memory consumption and slower performance.  

Solution  
Implement MinIO client for direct file uploads, storing only the file URLs instead of Base64 strings.  

Alternatives Considered  
- Keeping Base64 (inefficient for large files).  
- Using third-party cloud storage (MinIO offers a self-hosted solution).  

Approach  
1. Remove Base64 encoding logic.  
2. Integrate MinIO file upload for media messages.  
3. Store and retrieve file URLs efficiently.  

Additional Context  
This change will improve chat performance, reduce payload size, and optimize storage usage.  

---
### NOTE

1. We recently discovered a flaw in the code base where there are some XSS and CORS vulnerabilities. This occurs when the API and Admin servers run on different machines.
2. You will notice this if you configure your Admin app on your local machine to use the API running on https://test.talawa.io/graphql.
1. The errors in your browser will look like this:

    Image

1. As part of this issue you will need to ensure that the browser only interacts with the API through the Admin server and never with the API directly.

### PR DESCRIPTION
**Pull request by Nivedita-Chhokar**
**What kind of change does this PR introduce?**

Feature enhancement - Replaces Base64 encoding with MinIO client for chat attachments

**Issue Number:**

Fixes #3723 

**Snapshots/Videos:**

N/A

**If relevant, did you update the documentation?**

No

**Summary**

This PR replaces the Base64 storage approach for chat attachments with MinIO cloud storage to improve performance and reduce database size. The changes:

Remove Base64 encoding logic from file uploads
Integrate the existing MinIO client for file attachment handling
Store URLs to files instead of embedding large Base64 strings in messages

**Does this PR introduce a breaking change?**

No

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [ ] I have written tests for all new changes/features
- [ ] I have verified that test coverage meets or exceeds 95%
- [ ] I have run the test suite locally and all tests pass

**Other information**

None

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

- **New Features**
	- Introduced a streamlined file upload capability in the chat interface that enables direct attachment uploads to cloud storage.
	- Improved attachment management, automatically resetting the upload state upon message send or when the attachment is cleared.

- **Bug Fixes**
	- Enhanced error handling for file uploads, providing user notifications for upload failures.

- **Tests**
	- Added comprehensive unit tests for the `ChatRoom` component, covering various functionalities and edge cases related to messaging and file uploads, including rendering states and error handling.

### CODE REVIEW

**Comment by palisadoes** [comment]
1. Please make coderabbit.ai approves your work.
2. Make sure all tests pass and are valid.
3. Ensure the test code coverage for your patch reaches as close to **100%** as possible.
    - https://app.codecov.io/gh/PalisadoesFoundation/talawa-admin/pull/3823?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation
image

File: src/components/UserPortal/ChatRoom/ChatRoom.tsx [hunk 1]
```diff
@@ -203,13 +210,25 @@ export default function chatRoom(props: InterfaceChatRoomProps): JSX.Element {
...
+
+        const { fileUrl, objectName } = await uploadFileToMinio(
+          file,
+          organizationId,
+        );
+
+        setAttachment(fileUrl);
+        setAttachmentObjectName(objectName);
+      } catch (error) {
+        console.error('Error uploading file:', error);
```
Comment (varshith257) [inline src/components/UserPortal/ChatRoom/ChatRoom.tsx:230]: Add toast error to it as coderaabbitai suggested

**Review by varshith257 — COMMENTED** [review]
Also, can you attach a video of uploading file with this patch

File: src/components/UserPortal/ChatRoom/ChatRoom.tsx (hunk 1 above)
Comment (Nivedita-Chhokar) [reply src/components/UserPortal/ChatRoom/ChatRoom.tsx:230]: okay

**Comment by iamanishx** [comment]
@Nivedita-Chhokar mutation has been updated in the recent PR in api please check out

File: src/components/UserPortal/ChatRoom/ChatRoom.tsx [hunk 2]
```diff
@@ -203,13 +211,26 @@
...
+        toast.error('File upload failed. Please try again.');
+      }
     }
```
[the remark this answers is not in the document]
Comment (Nivedita-Chhokar) [reply src/components/UserPortal/ChatRoom/ChatRoom.tsx:234]: Thank you for the suggestion, but I believe there may be a misunderstanding. The TypeScript error indicates that uploadFileToMinio actually returns { fileUrl, objectName } rather than { fileHash, objectName }. Making the suggested change would likely cause errors since the property doesn't exist in the returned object.

File: src/components/UserPortal/ChatRoom/ChatRoom.spec.tsx [hunk 3]
```diff
@@ -5983,3 +5984,362 @@
...
+  let setAttachmentObjectName: jest.Mock;
+  let handleImageChange: (
+    e: React.ChangeEvent<HTMLInputElement>,
+  ) => Promise<void>;
+
+  beforeEach(() => {
+    vi.clearAllMocks();
+
+    setAttachment = vi.fn();
+    setAttachmentObjectName = vi.fn();
```
[the remark this answers is not in the document]
Comment (Nivedita-Chhokar) [reply src/components/UserPortal/ChatRoom/ChatRoom.spec.tsx:6139]: done

**Comment by palisadoes** [comment]
@iamanishx PTAL

**Comment by palisadoes** [comment]
1. Please make coderabbit.ai approves your work.
2. Resolve any conflicting files
3. Make sure all tests pass and are valid.
4. Ensure the test code coverage for your patch reaches as close to **100%** as possible.

**Comment by palisadoes** [comment]
1. Please make coderabbit.ai approves your work.
2. Resolve any conflicting files
3. Make sure all tests pass and are valid.
4. Ensure the test code coverage for your patch reaches as close to **100%** as possible.

**Comment by palisadoes** [comment]
Closing as abandoned
